src/vm/checker/typecheck/mod.rs

```diff
@@ -0,0 +1,404 @@
+                                              typed_result);
+        }
+
+        Ok(out_context)
+    }
+
+    fn get_function_type(&self, function_name: &str) -> Option<FunctionType> {
+        if let Some(function_type) = self.contract_context.get_function_type(function_name) {
+            Some(function_type.clone())
+        } else {
+            None
+        }
+    }
+
+    fn type_check_define_function(&mut self, function_expression: &[SymbolicExpression],
+                                  context: &TypingContext) -> CheckResult<(String, FunctionType)> {
+        if function_expression.len() != 3 {
+            return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(2, function_expression.len() - 1)))
+        }
+
+        self.type_map.set_type(&function_expression[0], no_type())?;
+        self.type_map.set_type(&function_expression[1], no_type())?;
+        // should we set the type of the subexpressions of the signature to no-type as well?
+
+        let signature = function_expression[1].match_list()
+            .ok_or(CheckError::new(CheckErrors::DefineFunctionBadSignature))?;
+        let body = &function_expression[2];
+
+        let (function_name, args) = signature.split_first()
+            .ok_or(CheckError::new(CheckErrors::VariadicNeedsOneArgument))?;
+        let function_name = function_name.match_atom()
+            .ok_or(CheckError::new(CheckErrors::BadFunctionName))?;
+        let mut args = parse_name_type_pairs(args)
+            .map_err(|_| { CheckError::new(CheckErrors::BadSyntaxBinding) })?;
+
+        let mut function_context = context.extend()?;
+        for (arg_name, arg_type) in args.iter() {
+            function_context.variable_types.insert(arg_name.clone(),
+                                                   arg_type.clone());
+        }
+
+
+        let return_type = self.type_check(body, &function_context)?;
+        let arg_types: Vec<TypeSignature> = args.drain(..)
+            .map(|(_, arg_type)| arg_type).collect();
+
+        Ok((function_name.to_string(), FunctionType::Fixed(arg_types, return_type)))
+    }
+
+    pub fn type_check_define_map(&mut self, map_expression: &[SymbolicExpression],
+                                 _context: &TypingContext) -> CheckResult<(String, (TypeSignature, TypeSignature))> {
+        if map_expression.len() != 4 {
+            return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(3, map_expression.len() - 2)))
+        }
+
+        self.type_map.set_type(&map_expression[0], no_type())?;
+        self.type_map.set_type(&map_expression[1], no_type())?;
+        self.type_map.set_type(&map_expression[2], no_type())?;
+        self.type_map.set_type(&map_expression[3], no_type())?;
+        // should we set the type of the subexpressions of the signature to no-type as well?
+
+        let map_name = map_expression[1].match_atom()
+            .ok_or(CheckError::new(CheckErrors::BadMapName))?;
+        let key_type = &map_expression[2];
```

Comment:
I'm having a hard time proving this to myself -- can I create a `map` whose `key_type` is a function?  I don't think I should be able to, but I haven't yet found the check that prevents this.

Reply:
The `key_type` and `value_type` variables will be filled when it tries to parse the TypedTuple definition. If you tried to pass a function there, the TypedTuple definition would be invalid, so it would raise an error in `parse_name_type_pair_list`, which would get mapped to `BadMapTypeDefinition`.